company.el

```diff
@@ -3609,27 +3707,32 @@ Delay is determined by `company-tooltip-idle-delay'."
   (let ((selection (or company-selection 0)))
     (let ((limit (window-body-width (minibuffer-window)))
           (len -1)
-          ;; Roll to selection.
           (candidates (nthcdr selection company-candidates))
-          (numbered (if company-show-numbers selection 99999))
+          (numbered (if company-show-quick-access selection 99999))
+          (qa-keys-len (length company-quick-access-keys))
           comp msg)
 
       (while candidates
-        (setq comp (company-reformat (company--clean-string (pop candidates)))
+        (setq comp (propertize
+                    (company-reformat (company--clean-string (pop candidates)))
+                    'face
+                    'company-echo)
               len (+ len 1 (length comp)))
-        (if (< numbered 10)
-            (progn
-              (cl-incf numbered)
-              (setq comp (propertize (format "%d: %s" (mod numbered 10) comp)
-                                     'face 'company-echo))
-              (cl-incf len 3)
-              ;; FIXME: Add support for the `match' backend action, and thus,
-              ;; non-prefix matches.
-              (add-text-properties 3 (+ 3 (string-width (or company-common "")))
-                                   '(face company-echo-common) comp))
-          (setq comp (propertize comp 'face 'company-echo))
-          (add-text-properties 0 (string-width (or company-common ""))
-                               '(face company-echo-common) comp))
+        (let ((beg 0)
+              (end (string-width (or company-common ""))))
+          (when (< numbered qa-keys-len)
+            (let ((qa-hint
+                   (format "%s: " (funcall
+                                   company-quick-access-hint-function
+                                   numbered))))
+              (setq beg (string-width qa-hint)
+                    end (+ beg (string-width (or company-common ""))))
```

Comment:
I noticed a duplicated evaluation here (already done on L3722), it'll be fixed with the next push.